### PR TITLE
*: Improve region forward

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Ref #5839.

### What is changed and how does it work?

Improve the region forwarding logic.

```commit-message
```

### Check List

Tests

- Manual test

### Release note

```release-note
None.
```

### CODE REVIEW

File: server/forward.go [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+	case tsDeadlineCh <- dl:
+	case <-forwardCtx.Done():
+		return nil, forwardCtx.Err()
+	}
+
+	start := time.Now()
+	resp, err := s.forwardTSORequest(forwardCtx, request, forwardStream)
+	close(done)
+	if err != nil {
+		if strings.Contains(err.Error(), errs.NotLeaderErr) {
```
Comment (CabinfeverB) [inline server/forward.go:207]: How about add a `handleErr` func?

**Review by CabinfeverB — APPROVED** [review]

File: server/forward.go (hunk 1 above)
Comment (rleungx) [reply server/forward.go:207]: I don't change the logic here, maybe we can keep it as it is.

File: server/forward.go [hunk 2]
```diff
@@ -0,0 +1,504 @@
...
+		// Successfully stored the connection we created.
+		return newConn, nil
+	}
+	// Loaded a connection created/stored by another goroutine, so close the one we created
+	// and return the one we loaded.
+	newConn.Close()
+	return conn.(*grpc.ClientConn), nil
+}
+
+func (s *GrpcServer) getForwardedHost(ctx, streamCtx context.Context, serviceName ...string) (forwardedHost string, err error) {
```
Comment (lhy1024) [inline server/forward.go:366]:
```suggestion
func (s *GrpcServer) getForwardedHost(ctx, streamCtx context.Context, serviceNames ...string) (forwardedHost string, err error) {
```

Comment (rleungx) [reply server/forward.go:366]: It's just an optional parameter in stead of plural.

**Review by lhy1024 — APPROVED** [review]

**Pull request merged**